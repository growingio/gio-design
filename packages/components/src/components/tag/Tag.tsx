import React from 'react';
import { Close } from '@gio-design/icons';
import classnames from 'classnames';
import { TagProps } from './interface';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';

export const isToggleClose = (closable = false, persistCloseIcon = false): boolean => closable && !persistCloseIcon;

const Tag: React.FC<TagProps & React.HTMLAttributes<HTMLSpanElement>> = (props: TagProps) => {
  const {
    children,
    type = 'normal',
    size = 'middle',
    status,
    color,
    closable,
    onClose,
    onClick,
    disabled,
    persistCloseIcon = true,
    className,
    customizePrefixCls,
    style,
  } = props;
  const prefix = usePrefixCls('tag', customizePrefixCls);

  return (
    <span
      className={classnames(
        prefix,
        classnames({ [`${prefix}-size-${size}`]: size }),
        classnames({ [`${prefix}-type-${type}`]: type }),
        classnames({ [`${prefix}-status-${status}`]: status }),
        classnames({ [`${prefix}-color-${color}`]: color }),
        classnames({ [`${prefix}-closable`]: closable }),
        classnames({ [`${prefix}-closable-toggle`]: isToggleClose(closable, persistCloseIcon) }),
        classnames({ [`${prefix}-closable-disabled`]: disabled }),
        className
      )}
      aria-hidden="true"
      onClick={onClick}
      style={style}
    >
      <span className={`${prefix}-label`}>{children}</span>
      {closable && !disabled ? <Close className={`${prefix}-closable-icon`} onClick={onClose} /> : null}
    </span>
  );
};

export default Tag;
