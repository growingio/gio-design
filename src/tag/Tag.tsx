import React from 'react';
import { CloseOutlined } from '@gio-design/icons';
import classnames from 'classnames';
import { usePrefixCls } from '@gio-design/utils';
import { TagProps } from './interface';

const Tag: React.FC<TagProps & React.HTMLAttributes<HTMLSpanElement>> = (props: TagProps) => {
  const {
    children,
    type = 'normal',
    size = 'middle',
    status,
    closable,
    onClose,
    onClick,
    disabled,
    className,
    customizePrefixCls,
    style,
  } = props;
  const prefix = usePrefixCls('tag-new', customizePrefixCls);

  return (
    <span
      className={classnames(
        prefix,
        classnames({ [`${prefix}-size-${size}`]: size }),
        classnames({ [`${prefix}-type-${type}`]: type }),
        classnames({ [`${prefix}-status-${status}`]: status }),
        classnames({ [`${prefix}-closable`]: closable }),
        classnames({ [`${prefix}-closable-disabled`]: disabled }),
        className
      )}
      aria-hidden="true"
      onClick={onClick}
      style={style}
    >
      <span className={`${prefix}-label`}>{children}</span>
      {closable && !disabled ? <CloseOutlined className={`${prefix}-closable-icon`} onClick={onClose} /> : null}
    </span>
  );
};

export default Tag;
