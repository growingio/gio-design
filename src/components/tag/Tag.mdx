import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Tag from './Tag';

# Tag 标签

<Subtitle>进行标记和分类的标签，作为已选内容的标签。</Subtitle>

## 参数说明

<ArgsTable of={Tag} />

## 代码演示

### 默认

基础标签。

[Example](/?path=/story/basic-components-tag--default)

### 可关闭标签

标签可关闭。

[Example](/?path=/story/basic-components-tag--closable)
