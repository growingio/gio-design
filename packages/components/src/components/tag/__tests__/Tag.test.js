import React from 'react';
import Tag from '../index';
import {
  isProrupt,
  isLarge,
  isPredifinedColor,
  isPredifinedStatus,
  isToggleClose,
  getTypeClass,
  getStatusClass,
  getColorClass,
  getDeleteToggleClass,
  getDisabledClass,
} from '../Tag';
import '@gio-design/components/es/components/Tag/style/index.css';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

describe('<Tag />', () => {
  it('renders <Tag /> components', () => {
    const tree = renderer.create(<Tag />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have "gio-tag" class', () => {
    const tree = shallow(<Tag />);
    expect(tree.hasClass('gio-tag')).toBe(true);
  });

  it('should have "test-custom-tag" class', () => {
    const tree = shallow(<Tag customizePrefixCls="test-custom-tag" />);
    expect(tree.hasClass('test-custom-tag')).toBe(true);
  });

  it('should have "gio-delete-toggle', () => {
    const tree = shallow(<Tag closable={true} persistCloseIcon={false} />);
    tree.simulate('hover');
    expect(tree.hasClass('gio-tag-delete-toggle')).toBe(true);
    expect(tree.exists('.gio-tag-close')).toBe(true);
  });

  it('should onClose call back', () => {
    const onClose = (e) => {
      expect(true).toBe(true);
    };
    const tree = shallow(<Tag closable={true} persistCloseIcon={true} onClose={onClose} />);
    tree.find('.gio-tag-close').simulate('click');
  });
});

describe('Tag condition functions', () => {
  it('isProrupt', () => {
    expect(isProrupt('prorupt', false)).toBe(true);
    expect(isProrupt('normal', true)).toBe(true);
  });
  it('isLarge', () => {
    expect(isLarge('large')).toBe(true);
  });
  it('isPredifinedColor', () => {
    expect(isPredifinedColor('beta')).toBe(true);
    expect(isPredifinedColor('new')).toBe(true);
    expect(isPredifinedColor('asdljalksdj')).toBe(false);
  });
  it('isPredifinedStatus', () => {
    expect(isPredifinedStatus('success')).toBe(true);
    expect(isPredifinedStatus('warning')).toBe(true);
    expect(isPredifinedStatus('asdljalksdj')).toBe(false);
  });
  it('isPersistClose', () => {
    expect(isToggleClose(true, true)).toBe(false);
    expect(isToggleClose(true, false)).toBe(true);
    expect(isToggleClose(false, true)).toBe(false);
    expect(isToggleClose(false, false)).toBe(false);
  });
});

describe('Tag conditional classes functions', () => {
  const prefix = 'gio-tag';
  it('getTypeClass', () => {
    expect(getTypeClass(prefix, 'normal')).toBe('');
    expect(getTypeClass(prefix, 'prorupt', true)).toBe('gio-tag-prorupt');
    expect(getTypeClass(prefix, 'large', true)).toBe('gio-tag-prorupt gio-tag-large');
  });
  it('getStatusClass', () => {
    expect(getStatusClass(prefix, 'success')).toBe('gio-tag-status-success');
  });
  it('getColorClass', () => {
    expect(getColorClass(prefix, 'beta')).toBe('gio-tag-color-beta');
  });
  it('getDeleteToggleClass', () => {
    expect(getDeleteToggleClass(prefix, true)).toBe('gio-tag-delete-toggle');
    expect(getDeleteToggleClass(prefix, false)).toBe('');
  });
  it('getDisabledClass', () => {
    expect(getDisabledClass(prefix, true)).toBe('gio-tag-disabled');
  });
});
