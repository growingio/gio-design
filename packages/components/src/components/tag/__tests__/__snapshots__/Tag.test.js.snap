// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tag /> renders <Tag /> components 1`] = `
<span
  className="gio-tag"
/>
`;
